conjure-lib/src/main/java/com/palantir/conjure/java/lib/internal/ConjureCollections.java

```diff
@@ -51,79 +53,82 @@ private ConjureCollections() {
+            return (List<T>) conjureDoubleList.asUnmodifiable();
+        } else if (list instanceof ConjureSafeLongList conjureSafeLongList) {
+            return (List<T>) conjureSafeLongList.asUnmodifiable();
         } else {
             // Otherwise use the JDK types
             return Collections.unmodifiableList(list);
         }
     }
 
-    @SuppressWarnings("unchecked")
     public static <T> void addAll(Collection<T> addTo, Iterable<? extends T> elementsToAdd) {
-        Preconditions.checkNotNull(elementsToAdd, "elementsToAdd cannot be null");
-        if (elementsToAdd instanceof Collection) {
+        if (elementsToAdd instanceof Collection<? extends T> collection) {
             // This special-casing allows us to take advantage of the more performant
             // ArrayList#addAll method which does a single System.arraycopy.
-            addTo.addAll((Collection<T>) elementsToAdd);
+            addTo.addAll(collection);
         } else {
-            for (T element : elementsToAdd) {
-                addTo.add(element);
-            }
+            Preconditions.checkNotNull(elementsToAdd, "elementsToAdd cannot be null")
+                    .forEach(addTo::add);
         }
     }
 
-    @SuppressWarnings("unchecked")
     public static <T> void addAllAndCheckNonNull(Collection<T> addTo, Iterable<? extends T> elementsToAdd) {
-        Preconditions.checkNotNull(elementsToAdd, "elementsToAdd cannot be null");
-        // If we know the number of elements we are adding and the addTo Collection is an ArrayList, we can eagerly
-        // resize it to only do one grow() of the array.
-        if (elementsToAdd instanceof Collection) {
-            Collection<T> collectionElementsToAdd = (Collection<T>) elementsToAdd;
-            if (addTo instanceof ArrayList) {
-                ((ArrayList<T>) addTo).ensureCapacity(collectionElementsToAdd.size() + addTo.size());
-            }
-        }
-        for (T element : elementsToAdd) {
-            Preconditions.checkNotNull(element, "elementsToAdd cannot contain null elements");
-            addTo.add(element);
+        if (elementsToAdd instanceof Collection<? extends T> collection) {
+            // Some collections such as ArrayList support bulk addAll optimizations to avoid repeated resizing.
+            addTo.addAll(new AbstractCollection<>() {
+                @Override
+                public Iterator<T> iterator() {
+                    return new NonNullIterator<>(collection.iterator());
+                }
+
+                @Override
+                public Object[] toArray() {
+                    Object[] array = collection.toArray();
+                    for (Object element : array) {
+                        checkNotNullElement(element);
+                    }
+                    return array;
+                }
+
+                @Override
+                public int size() {
+                    return collection.size();
+                }
```

Comment:
In an ideal world I think we'd implement this within specialized non-null collections to avoid the new anonymous wrapper allocation on a per-call basis. Getting there would be a little tricky, because this API is fairly specialized, however it's in the `internal` package so we can make some assumptions about how it's used.
Probably not a substantial optimization.

Reply:
I suppose another option is to make the assumption that `toArray` will be used, and explicitly handle that ourselves:
```java
T[] values = elementsToAdd.toArray();
addTo.ensureCapacity[IfPossible](values.length);
for (T element : values) { // no more iterator allocation because we're iterating over an array
    Preconditions.checkNotNull(element, "elementsToAdd cannot contain null elements");
    addTo.add(element);
}
```